### PR TITLE
adding MVVM pattern scripts that get data from a remote server

### PR DESCRIPTION
**Pull request by Faballen**
Implemented the MVVM pattern in order to get data from a remote server
used Dagger Hilt and Retrofit

### CODE REVIEW

File: Fab-Rapptr-Assessment/app/build.gradle [hunk 1]
```diff
@@ -46,4 +48,20 @@ dependencies {
...
+    // Retrofit
+    implementation "com.squareup.retrofit2:retrofit:2.9.0"
+    // retrofit with gson converter
+    implementation 'com.squareup.retrofit2:converter-gson:2.9.0'
+    // Retrofit with Scalar Converter
+    implementation "com.squareup.retrofit2:converter-scalars:2.9.0"
+    // dagger dependency injection
+    implementation("com.google.dagger:hilt-android:2.44")
+    kapt("com.google.dagger:hilt-android-compiler:2.44")
+    // kotlin coroutine
```
Comment (KyleMcB) [inline Fab-Rapptr-Assessment/app/build.gradle:60]: this comment is not above the kotlin coroutine dependency. That is viewModel kotin extensions

File: Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/model/ChatLogMessageModel.kt [hunk 2]
```diff
@@ -1,8 +1,10 @@
 package com.example.fabrapptrassessment.chat.model
 
+import com.google.gson.annotations.SerializedName
```
Comment (KyleMcB) [inline Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/model/ChatLogMessageModel.kt:3]: This is weird. I don't see gson in the build file. Does the app build?

File: Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/network/ChatApiService.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.example.fabrapptrassessment.chat.network
+
+import com.example.fabrapptrassessment.chat.model.ChatLogMessageModel
+import retrofit2.Call
+import retrofit2.http.GET
+
+interface ChatApiService {
+    /**
```
Comment (KyleMcB) [inline Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/network/ChatApiService.kt:8]: I'm impressed you wrote some documentation like this. Great Work. There is a semi standard comment documentation format known as JavaDoc. I use this website a lot. [javaDoc cheatsheet](https://binfalse.de/2015/10/05/javadoc-cheats-sheet/)

File: Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/repository/ChatRepository.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+
+@Module
+@InstallIn(SingletonComponent::class)
+class ChatRepository @Inject constructor( private val  chatApiService: ChatApiService) {
+    // mutable live data tha can change withing the scope of the viewModel
+    private var _chat = MutableLiveData<List<ChatLogMessageModel>>()
+    val chat : LiveData<List<ChatLogMessageModel>> = _chat // view will observe the mutableLiveData
+
+    fun getListChat () { // An invocation of a Retrofit method that sends a request to a webserver and returns a response.
```
Comment (KyleMcB) [inline Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/repository/ChatRepository.kt:24]: This function could use some work. First, there is a return statement inside a try and not the catch. Does that compile? I suggest making a sealed interface and returning that here something like this
```Kotlin 
sealed interface ListChatResult {
  data class Success(val data: LiveData<List<ChatLogMessageModel>>) : ListChatResult
  object Failure : ListChatResult
}
fun getListChat() : ListChatResult {
  return try {
    chatApiService.getChat().enqueue(object : Callback<List<ChatLogMessageModel>> {
                override fun onResponse(
                    call: Call<List<ChatLogMessageModel>>,
                    response: Response<List<ChatLogMessageModel>>
                ) {
                    Log.d("response", "data : ${response.body()}")
                    _chat.postValue(response.body())
                }

                override fun onFailure(call: Call<List<ChatLogMessageModel>>, t: Throwable) {
                    Log.e("failure", "Failed to fetch data")
                    t.printStackTrace()
                }
            })
    _chat
  } catch (e : Exception) {
   ListChatResult.Failure 
  }
}
```

**Review by KyleMcB — APPROVED** [review]
nice work. looking forward to the next one

File: Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/model/ChatLogMessageModel.kt (hunk 2 above)
Comment (Faballen) [reply Fab-Rapptr-Assessment/app/src/main/java/com/example/fabrapptrassessment/chat/model/ChatLogMessageModel.kt:3]: Its there at the bottom. i added it so the variable in data class could be mapped to the JSON attribute

**Pull request merged**